File: tools/clod-poc/tools/qa-baseline.ts
import { resolve } from "node:path";
import { loadUnifiedRegistry } from "../src/qa/unified/manifest.js";
import { updateQaBaselines } from "../src/qa/unified/baseline.js";

const REPOSITORY_ROOT = resolve(import.meta.dirname, "../../..");
const VISUAL = resolve(REPOSITORY_ROOT, "validation/manifests/visual-regression.yaml");
const PERFORMANCE = resolve(REPOSITORY_ROOT, "validation/manifests/performance-regression.yaml");
const LEGACY = resolve(REPOSITORY_ROOT, "validation/manifests/legacy-id-map.yaml");

interface Args { runRoot?: string; scenes: string[]; approve: boolean; allowCi: boolean }

function main(): void {
  const args = parseArgs(process.argv.slice(2));
  if (!args.runRoot) throw new Error("--run-root is required");
  const registry = loadUnifiedRegistry({ visual: VISUAL, performance: PERFORMANCE, legacyMap: LEGACY });
  const authorities = updateQaBaselines(registry, {
    repositoryRoot: REPOSITORY_ROOT,
    runRoot: resolve(REPOSITORY_ROOT, args.runRoot),
    visualManifest: VISUAL,
    performanceManifest: PERFORMANCE,
    sceneIds: args.scenes,
    approve: args.approve,
    allowCi: args.allowCi,
  });
  console.log(`[qa-baseline] promoted ${authorities.length} baselines to version ${authorities[0]?.baseline_version ?? registry.baselineVersion}`);
}

function parseArgs(argv: readonly string[]): Args {
  const args: Args = { scenes: [], approve: false, allowCi: false };
  for (let index = 0; index < argv.length; index++) {
    const arg = argv[index];
    const value = argv[index + 1];
    if (arg === "--run-root" && value) { args.runRoot = value; index++; }
    else if (arg === "--scene" && value) { args.scenes.push(value); index++; }
    else if (arg === "--approve") args.approve = true;
    else if (arg === "--allow-ci") args.allowCi = true;
    else throw new Error(`unknown or incomplete argument: ${arg}`);
  }
  return args;
}

try { main(); }
catch (error) {
  console.error("[qa-baseline] error:", error instanceof Error ? error.message : error);
  process.exit(1);
}

File: src/bin/qa_baseline.rs
use std::path::PathBuf;
use std::process::ExitCode;

use clap::Parser;
use voxel_builder::diagnostics::qa::unified::baseline::{BaselineOptions, promote_baselines};
use voxel_builder::diagnostics::qa::unified::manifest::load_registry_with_legacy;

#[derive(Parser, Debug)]
#[command(about = "Promote an authoritative unified QA run to canonical baselines", version)]
struct Args {
    #[arg(long)]
    run_root: PathBuf,
    #[arg(long)]
    scene: Vec<String>,
    #[arg(long)]
    approve: bool,
    #[arg(long)]
    allow_ci: bool,
    #[arg(long, default_value = "validation/manifests/visual-regression.yaml")]
    visual: PathBuf,
    #[arg(long, default_value = "validation/manifests/performance-regression.yaml")]
    performance: PathBuf,
    #[arg(long, default_value = "validation/manifests/legacy-id-map.yaml")]
    legacy_map: PathBuf,
}

fn main() -> ExitCode {
    match run() {
        Ok(count) => { println!("[qa-baseline] promoted {count} canonical baselines"); ExitCode::SUCCESS }
        Err(error) => { eprintln!("[qa-baseline] error: {error}"); ExitCode::from(1) }
    }
}

fn run() -> Result<usize, String> {
    let args = Args::parse();
    let repository_root = std::env::current_dir().map_err(|error| error.to_string())?;
    let registry = load_registry_with_legacy(&args.visual, &args.performance, Some(&args.legacy_map))
        .map_err(|error| error.to_string())?;
    let authorities = promote_baselines(
        &registry,
        &BaselineOptions {
            repository_root,
            run_root: args.run_root,
            visual_manifest: args.visual,
            performance_manifest: args.performance,
            scene_ids: args.scene,
            approve: args.approve,
            allow_ci: args.allow_ci,
        },
    ).map_err(|error| error.to_string())?;
    Ok(authorities.len())
}
